Offpeak in db option

jaykorean opened the pull request:
# Summary

RocksDB's primary function is to facilitate read and write operations. Compactions, while essential for minimizing read amplifications and optimizing storage, can sometimes compete with these primary tasks. Especially during periods of high read/write traffic, it's vital to ensure that primary operations receive priority, avoiding any potential disruptions or slowdowns. Conversely, during off-peak times when traffic is minimal, it's an opportune moment to tackle low-priority tasks like TTL based compactions, optimizing resource usage.

In this PR, we are incorporating the concept of off-peak time into RocksDB by introducing `daily_offpeak_time_utc` within the DBOptions. This setting is formatted as "HH:mm-HH:mm" where the first one before "-" is the start time and the second one is the end time, inclusive. It will be later used for resource optimization in subsequent PRs.

# Test Plan

- New Unit Test Added - `DBOptionsTest::OffPeakTimes`
- Existing Unit Test Updated - `OptionsTest`, `OptionsSettableTest`

jaykorean commented on options/db_options.cc:
```diff
@@ -1056,6 +1070,34 @@ void MutableDBOptions::Dump(Logger* log) const {
+                   daily_offpeak_end_time_utc.c_str());
+}
+
+bool MutableDBOptions::IsNowOffPeak(SystemClock* clock) const {
```
Not sure if this is the best place for this util function. I can create a new util, maybe call it `time_util` or something else. Please let me know.

Also, I am still a C++ n00b. If there's a better way to do what I did here, please let me know.

jaykorean commented on include/rocksdb/options.h:
```diff
@@ -1427,6 +1427,21 @@ struct DBOptions {
+  //
+  // Time of the day in UTC. Format - HH:mm (e.g. 00:00 - 23:59)
+  // Default: Empty String (No notion of peak/offpeak)
+  std::string daily_offpeak_start_time_utc = "";
```
Another way to do this is use `std::optional<std::string>`

jaykorean commented on include/rocksdb/options.h:
```diff
@@ -1427,6 +1427,21 @@ struct DBOptions {
   // inconsistency, e.g. deleted old data become visible again, etc.
   bool enforce_single_del_contracts = true;
+
+  // EXPERIMENTAL
```
I think it's safe to say that it's experimental for now, but please let me know if anyone feel this is not necessary.

pdillinger commented:
Thanks, review in progress :)

pdillinger commented on options/db_options.cc:
```diff
@@ -1056,6 +1070,34 @@ void MutableDBOptions::Dump(Logger* log) const {
+    time_t now_unix_time = static_cast<time_t>(now);
+    // extract hour and minute from unix_time_now in HH:mm format
+    char now_utc[6];
+    std::strftime(now_utc, sizeof(now_utc), "%R", std::gmtime(&now_unix_time));
```
Leaning on standard APIs is generally a good idea, but I worry that we're bringing in extra complexity that we might not fully understand and foresee corner cases of. We compile on a lot of different systems and locales, and we have been hit by variances in std api behavior before.

Note that unix time 0 is midnight UTC and that generalizes such that any unix time % 86400 is the number of seconds since midnight. I think the IsNowOffPeak determination should be made by converting to these "seconds since midnight" values. String comparison for time comparison is somewhat fragile, as it relies on zero padding and (here) implicit conversion of now_utc from char* to string. "Seconds since midnight" makes it easier to expand the format of our allowed time strings in the future, without really affecting this function (assuming we have a parse function used by both this function and ValidateOptions).

jaykorean replied:
Yes, I sort of realized that "seconds since midnight" would work after writing the test. Because what I did in the test to set the time was basically `SetCurrentTime(now_utc_hour * 3600 + now_utc_minute * 60)`). Just thinking backwards, this would work for checking the time as well. Let me make the change.

pdillinger commented on db/db_options_test.cc:
```diff
@@ -1033,6 +1034,76 @@ TEST_F(DBOptionsTest, SetFIFOCompactionOptions) {
+  options.daily_offpeak_start_time_utc = "06:30";
+  options.daily_offpeak_end_time_utc = "11:30";
+  verify_is_now_offpeak(5, 30, false);
+  verify_is_now_offpeak(6, 30, true);
```
Good tests, though we should test that 1 minute can make the difference, and (ideally) no digits are dropped / ignored.

jaykorean replied:
Let me also add a test to set this with SetDBOptions API

pdillinger reviewed (commented):
Some recommended changes / enhancements. Generally looks good.

(as discussed offline...) As much as I hate to put the timezone problem on the user, I agree that's better in the long run than making the meaning of our options depend on the configured time zone, which entangles us in potential complexity and uncertainty.

pdillinger commented:
Also, isn't there roughly enough here to include a test that uses SetDBOptions to dynamically change the offpeak times?

jaykorean commented on db/db_impl/db_impl_open.cc:
```diff
@@ -291,6 +292,29 @@ Status DBImpl::ValidateOptions(const DBOptions& db_options) {
+      db_options.daily_offpeak_end_time_utc != "") {
+    if (db_options.daily_offpeak_start_time_utc == "" ||
+        db_options.daily_offpeak_end_time_utc == "") {
+      return Status::InvalidArgument(
```
I like the idea of putting them into a single option, too. Let me make the change.

mrambacher replied:
I agree that a single option is probably better.  Would it also then be possible to define/document a format that allows multiple offpeak times?  For example, from 1AM-5AM and 7PM-11PM.  How this "vector" is separated is left up to you to define.

jaykorean replied:
Thanks for your comment. Initially, we considered the possibility of having multiple peak and off-peak periods throughout the day. However, after an internal discussion, we decided to opt for a single off-peak time period for the following reasons:

1. Managing too many peak and off-peak periods could introduce unnecessary complexity. If such a feature needs to be supported in the future, it might be more practical to provide an API like `SetPeak(bool peak)` for users to explicitly designate peak times at the start or end of those periods.

2. The primary objective of defining this time period is to enhance resource utilization by preprocessing low-priority tasks, such as TTL-based compactions, during off-peak hours. This approach minimizes the chances of these tasks competing for resources with read and write operations during peak hours. Therefore, establishing a single off-peak time frame aligns better with this goal compared to having multiple peak times throughout the day. This approach is similar to what HBase does (http://bigdatainourpalm.blogspot.com/2017/09/how-to-control-time-during-which-hbase.html).

jaykorean commented:
Thanks for the review @pdillinger. Addressed the feedback as I made the following changes
- `daily_offpeak_start_time_utc` and `daily_offpeak_end_time_utc` are now combined into `daily_offpeak_time_utc` requiring format `HH:mm-HH:mm`
- regex removed. Instead `ParseTimeStringToSeconds()` is introduced to `string_util.h` which returns seconds since midnight from "HH:mm" formatted input
- Instead of string comparison, `IsNowOffPeak()` is comparing seconds since midnight.
- Extended `DBOptionsTest::OffPeakTimes` to check against Get/SetDBOptions. The test now also opens db with mock clock and offpeak option set, then uses mock sleep to see if `isNowOffPeak()` changes accordingly.

PTAL when you get a chance

mrambacher commented on options/db_options.cc:
```diff
@@ -1056,6 +1065,33 @@ void MutableDBOptions::Dump(Logger* log) const {
+  }
+  int64_t now;
+  if (clock->GetCurrentTime(&now).ok()) {
+    int since_midnight_seconds = static_cast<int>(now % 86400);
```
Please make the 86400 into a constexpr

mrambacher commented on options/db_options.cc:
```diff
@@ -6,6 +6,9 @@
 #include "options/db_options.h"
 
 #include <cinttypes>
+#include <cstdint>
+#include <ctime>
+#include <iomanip>
```
Are these includes still required?

jaykorean replied:
Thanks for catching this. They aren't needed anymore. I was expecting our internal linter to clean them up, but I guess it doesn't work that way. Let me fix that.

mrambacher commented on options/db_options.cc:
```diff
@@ -1011,7 +1019,8 @@ MutableDBOptions::MutableDBOptions(const DBOptions& options)
       compaction_readahead_size(options.compaction_readahead_size),
-      max_background_flushes(options.max_background_flushes) {}
+      max_background_flushes(options.max_background_flushes),
+      daily_offpeak_time_utc(options.daily_offpeak_time_utc) {}
```
Since this time is really only HH:mm, should it just be "daily_offpeak_time" rather than "daily_offpeak_time_utc"?

jaykorean replied:
We want to explicitly say that it needs to be in UTC. Many of the internal configurations are often in local timezone (Pacific). Without `_utc` suffix, users may wrongfully assume that the value should be in the  local timezone, not UTC.

For example, let's say the server is running in PDT and the user wants to set the offpeak period from 11:30PM to 4:30AM next day. If they set `23:30-04:30`, `IsNowOffPeak()` will return true from 4:30PM to 9:30PM PDT which is not what the user intended. What they actually want to set is `06:30-11:30` which is in UTC.

jaykorean replied:
And yes, it's an unfortunate extra burden for users to figure out the conversion and set the correct value in UTC. Per internal discussion, we concluded that fixing the value in UTC is better than our options dependent on the configured time zone... (see also @pdillinger 's comment - https://github.com/facebook/rocksdb/pull/11893#pullrequestreview-1647346844)

jaykorean commented on db/db_options_test.cc:
```diff
@@ -1033,6 +1034,137 @@ TEST_F(DBOptionsTest, SetFIFOCompactionOptions) {
+
+  std::vector<std::string> valid_cases = {
+      "",  // Not enabled. Valid case
+      "06:30-11:30",
```
That's added in line 1104 as part of `verify_is_now_offpeak` which also checks for validity of the range.

jaykorean commented on db/db_options_test.cc:
```diff
@@ -1033,6 +1034,137 @@ TEST_F(DBOptionsTest, SetFIFOCompactionOptions) {
+  std::vector<std::string> valid_cases = {
+      "",  // Not enabled. Valid case
+      "06:30-11:30",
+      "06:30-23:30",
```
Ah, my first version of this PR made it invalid when the start time and end time were separate values, but the validation got removed as I made the change. 

`IsNowOffPeak()` is always going to return `false` in that case and nothing bad really will happen, but still, I think it's safe to add to avoid user confusion.

jaykorean replied:
Actually, after another thought, I will leave the same values as valid and it will work as the entire 24 hours are considered offpeak.

Currently, there's no way for users to set the entire 24 hours to be offpeak. There's always going to be 1 minute gap.

jaykorean replied:
Yeah, that was a bug and had to be fixed, but I ended up not allowing start_time == end_time and introduced a new boolean option for "always offpeak" Thanks again for the review.

jaykorean commented on db/db_options_test.cc:
```diff
@@ -1033,6 +1034,137 @@ TEST_F(DBOptionsTest, SetFIFOCompactionOptions) {
+      "06:30-11:30",
+      "06:30-23:30",
+      "13:30-14:30",
+      "23:30-01:15",  // From 11:30PM to 1:15AM next day. Valid case
```
Like I mentioned in the earlier comment, we've decided to have multiple offpeak periods to avoid unnecessary complexity. I think our API is clear enough for user to understand the overnight timespan.

If there's a strong need of multiple offpeak periods later, we can add support for different format as you suggested (a string with multiple dates and a well-known separator)

jaykorean commented on db/db_impl/db_impl_open.cc:
```diff
@@ -291,6 +292,15 @@ Status DBImpl::ValidateOptions(const DBOptions& db_options) {
   }
 
+  if (db_options.daily_offpeak_time_utc != "") {
+    auto split = StringSplit(db_options.daily_offpeak_time_utc, '-');
+    if (split.size() != 2 || ParseTimeStringToSeconds(split[0]) < 0 ||
+        ParseTimeStringToSeconds(split[1]) < 0) {
+      return Status::InvalidArgument(
```
Hm.. do you mean we are not validating the option when it's changed by `SetDBOptions()`?

`SetDBOptions()` do call `ValidateOptions()` and fail with an invalid option and this was verified in the test that I wrote.

mrambacher replied:
No, I mean that if you call ConfigureDBOptionsFromString("offpeak=bad"), the Configure call would fail because the value is not valid.

jaykorean replied:
IIUC you mean moving the validation / verification logic to option parser? Please feel free to put up a PR for the change after I merge this PR. I will be more than happy to review and work with you on this.

pdillinger commented on util/string_util.cc:
```diff
@@ -437,6 +437,45 @@ bool SerializeIntVector(const std::vector<int>& vec, std::string* value) {
+  char colon;
+
+  std::istringstream stream(value);
+  stream >> hours >> colon >> minutes;
```
Not a big deal, but cases like "0:1-2:000000000000000042" are considered valid. It doesn't bother me but worth noting in the implementation and/or test if not fixing.

jaykorean replied:
That's a weird thing for users to try but sure, let me add it to the test. As long as it's a valid number format that we can parse, there's no reason to block it I think.

pdillinger commented on db/db_options_test.cc:
```diff
@@ -1033,6 +1034,140 @@ TEST_F(DBOptionsTest, SetFIFOCompactionOptions) {
+  verify_is_now_offpeak(4, 30, true);
+  verify_is_now_offpeak(4, 31, false);
+
+  // entire day is offpeak (weird use case, but valid)
```
Except one minute, right? ;)

jaykorean replied:
Ah you are right. Let me fix the comment and add a special case in the test.

pdillinger commented on db/db_options_test.cc:
```diff
@@ -1033,6 +1034,140 @@ TEST_F(DBOptionsTest, SetFIFOCompactionOptions) {
+  options.daily_offpeak_time_utc = "23:30-04:30";
+  auto mock_clock = std::make_shared<MockSystemClock>(env_->GetSystemClock());
+  auto mock_env = std::make_unique<CompositeEnvWrapper>(env_, mock_clock);
+  mock_clock->SetCurrentTime(now_hour * 3600 + now_minute * 60);
```
Could improve the test slightly by adding some large multiple of 86400

pdillinger reviewed (approved):
Looks great to me!

mrambacher reviewed (commented):
Minor changes suggested

jaykorean commented:
Thanks @mrambacher for the review and suggestions! Just addressed them. PTAL and let me know.

mrambacher commented on db/db_options_test.cc:
```diff
@@ -1033,6 +1034,159 @@ TEST_F(DBOptionsTest, SetFIFOCompactionOptions) {
+  verify_is_now_offpeak(true, 4, 30);
+  verify_is_now_offpeak(false, 4, 31);
+
+  // There's one minute gap from 11:59PM to midnight
+  options.daily_offpeak_time_utc = "00:00-23:59";
+  verify_is_now_offpeak(true, 0, 0);
+  verify_is_now_offpeak(true, 12, 00);
+  verify_is_now_offpeak(true, 23, 59);
+  verify_is_now_offpeak(false, 23, 59, 1);
```
Can you please add a case to verify that the entire day can be off-peak if start==end?

jaykorean replied:
Line 1073 has it :)

jaykorean replied:
I was able to catch a bug by writing the test. Thanks! Fix and test are added in #11911

mrambacher reviewed (commented):
Generally looks good.  I cannot approve this.  Suggest adding a test where the entire day is offpeak to verify the time works.
